### ISSUE TITLE
hit animation bug

### ISSUE THREAD

**Issue by pengyz** [wheybags/freeablo#255]
First of all,I'm very sorry for my poor english.

I found that we always got a strange animation display when attacking the monster.

1. player will change his location when we enter the hit animation. I don't know why... maybe our monster should have an collision body.Maybe we caculate the direction with a wrong point for monster.
2. monster is hitted in a wrong time. Beacuse we just change the monster's hit animation at the begin of player's attack animation. Consider use a message bus and fire hit message at the end of player's attack animation.

How to attack a monster? see this code:
```C++
bool Player::attack(Actor *enemy)
    {
        if(enemy->isDead() && enemy->mStats != nullptr)
            return false;
        isAttacking = true;
        Engine::ThreadManager::get()->playSound(FALevelGen::chooseOne({"sfx/misc/swing2.wav", "sfx/misc/swing.wav"}));
        enemy->takeDamage((uint32_t)mStats->getMeleeDamage());//call takeDamage and immediate change the monster's animation.
        if(enemy->getCurrentHP() <= 0)
            enemy->die();
        setAnimation(AnimState::attack, false);
        mAnimPlaying = true;
        return true;
    }

void Actor::takeDamage(double amount)
    {
        mStats->takeDamage(static_cast<int32_t> (amount));
        if (!(mStats->getCurrentHP() <= 0))
        {
            Engine::ThreadManager::get()->playSound(getHitWav());
            setAnimation(AnimState::hit);//yes we display hit animation.
            mAnimPlaying = true;
        }
        else
            mAnimPlaying = false;
    }`
```
Consider more complex condition... Maybe our attack is missed,so we must change this animation logic.

**Comment by wheybags** [comment]
The reason is actually an unfortunate side effect of the original game assets.
CEL files are encoded in reverse scanline order, starting from the bottom right, unlike, well, every sensible format ever.
We render sprites by specifying a pixel position for the top left of the sprite, so when we change animation, the "origin" is not always the same, since the origin is at the bottom right not top left.
I'm not quite sure yet how we should deal with this problem. I don't want modders to have to deal with making sprites whose origin is bottom right, that seems mad, so maybe some sort of on-the-fly grooming of CEL assets to rework their origin?

As for combat triggering hits incorrectly, yeah, a large chunk of the combat code needs to be redone, it doesn't follow the actual formulae from diablo properly either.

**Comment by pengyz** [comment]
Thanks for your answer. I understand it now.

It seems a rather bad news...
Should we just fix it by use the bottom right for diablo's original sprite?

**Comment by mewmew** [comment]
@wheybags Would this issue be solved if #23 was ever implemented?
